src/main/java/com/github/packageurl/PackageURL.java

```diff
@@ -441,24 +442,49 @@ private String percentEncode(final String input) {
     }
 
     private static String uriEncode(String source, Charset charset) {
-        if (source == null || source.length() == 0) {
+        if (source == null || source.isEmpty()) {
             return source;
         }
 
-        StringBuilder builder = new StringBuilder();
-        for (byte b : source.getBytes(charset)) {
+        byte[] bytes = source.getBytes(charset);
+        int length = bytes.length;
+        int pos = indexOfFirstUnreservedChar(bytes);
+
+        if (pos == -1) {
+            return source;
+        }
+
+        StringBuilder builder = new StringBuilder(source.substring(0, pos));
+
+        for (int i = pos; i < length; i++) {
+            byte b = bytes[i];
+
             if (isUnreserved(b)) {
                 builder.append((char) b);
-            }
-            else {
-                // Substitution: A '%' followed by the hexadecimal representation of the ASCII value of the replaced character
+            } else {
                 builder.append('%');
-                builder.append(Integer.toHexString(b).toUpperCase());
+                builder.append(Character.toUpperCase(Character.forDigit((b >> 4) & 0xF, 16)));
+                builder.append(Character.toUpperCase(Character.forDigit(b & 0xF, 16)));
             }
         }
+
         return builder.toString();
     }
 
+    private static int indexOfFirstUnreservedChar(final byte[] bytes) {
```

Comment:
The name of this method is not accurate. `isUnreserved` returns whether the character is unreserved in a URI. This method returns the index of the first _not_ unreserved character, which is not the same as the index of the first reserved character because reserved is not the compliment of unreserved. I'm not sure what to call this that doesn't result in a double negative.

Reply:
How about `indexOfFirstUnsafeChar`?